build.gradle.kts

```diff
@@ -104,7 +104,7 @@ dependencies {
      * - we stop using testcontainers in docker-infrastructure
      * - we stop using docker-infrastructure
      */
-    testCompile("org.testcontainers:testcontainers:1.14.3")
+    testCompile("org.testcontainers:testcontainers:1.15.0")
```

Comment:
Hey :) Just curious - what does "they shade docker-java-core properly" mean? Also, have you re-evaluated it in 1.15.0? We no longer shade `docker-java-api` and only shade `docker-java-core`, and, I believe... properly :D

Reply:
Hi @bsideup  👋 
I commented this in 4b242c3db69a9887b069159d713319ce266edfa3 (in 1.14.3 times). The commit has some context, but I'll add more. When we depended on `docker-java-core`, we had two copies of `com.github.dockerjava.core.DockerClientImpl`. We were at the mercy of the classpath ordering. Also, the copies were binary-incompatible, e.g. 29c29fc6d66e4376b0b8ac3c127503211a69cc03.

By "proper shading" I meant copying the classes into a different package name, to avoid such conflicts.
Original [Maven "shading" definition](https://maven.apache.org/plugins/maven-shade-plugin/index.html):

> shade - i.e. rename - the packages of some of the dependencies

That said, I see 1.15.0 is indeed shading it properly: `org.testcontainers.shaded.com.github.dockerjava.core.DockerClientImpl`. It no longer clashes, so I can remove the workaround. Thanks! 🙌

I appreciate the work on sorting out the transitive dependency conflicts 🙏

Reply:
@bsideup PS. it would be even nicer if we could avoid shading.

The lib consumer can pick the `docker-java-core` they need. If it's incompatible with `testcontainers` transitive dep, the conflict is discoverable (e.g. by Gradle `failOnVersionConflict` or testing) and fixable by the consumer.

We could help with dep conflicts even futher. If the dep is narrow (single version), then it would produce bogus conflicts. We could appeal to `docker-java` maintainers to express (and protect) compatibility via SemVer. Then TC could declare compatibility with a range of `docker-java` versions. Then each detected conflict would be legit.

We could help lib consumers avoid classpath hell by taking advantage of the POM dep declarations. But maybe shading is necessary in this case. Just sharing my utopian dream 😉

Reply:
> We could appeal to docker-java maintainers to express (and protect) compatibility via SemVer


You're talking to _the_ maintainer 😂

The problem is that I couldn't get rid of unnecessary dependencies in -core in time, and historically we didn't want to have Guava on Testcontainers's list of dependencies, among a few others

We did a major effort migrating to non-shaded docker-java-api but -core will take a bit of time 😀

Reply:
> You're talking to the maintainer 😂

Aaah, right 😃 apologies, I made this mistake before 😅 

> We did a major effort migrating to non-shaded docker-java-api but -core will take a bit of time 😀

Glad to hear.
Guava is a real troublemaker for sure. I found its value is pretty shallow and each call is quite easy to weed out. Hopefully it's just a matter of usage volume rather than depth.

Reply:
I am planning to remove the Guava usages (as well as a few others that are not necessary) in future, so it is only a matter of time :)